specs/altair/validator.md

```diff
@@ -0,0 +1,396 @@
+
+This document is currently illustrative for early Altair testnets and some parts are subject to change. 
+In particular, the sync committee aggregation scheme may change to reduce resource consumption and better leverage the subcommittee aggregates defined in the state. 
+This change could touch the format of the `SyncCommitteeContribution`, the logic of the proposer to assemble the final `SyncAggregate`, and the derivation of the sync committee `subnet_id`. 
+Other changes may occur pending initial implementation and profiling for Altair testnets.
+
+## Constants
+
+### Misc
+
+| Name | Value | Unit |
+| - | - | :-: |
+| `TARGET_AGGREGATORS_PER_SYNC_SUBCOMMITTEE` | `2**2` (= 4) | validators |
+| `SYNC_COMMITTEE_SUBNET_COUNT` | `8` | The number of sync committee subnets used in the gossipsub aggregation protocol. |
+
+## Containers
+
+### `SyncCommitteeSignature`
+
+```python
+class SyncCommitteeSignature(Container):
+    # Slot to which this contribution pertains
+    slot: Slot
+    # Block root for this signature
+    beacon_block_root: Root
+    # Index of the validator that produced this signature
+    validator_index: ValidatorIndex
+    # Signature by the validator over the block root of `slot`
+    signature: BLSSignature
+```
+
+### `SyncCommitteeContribution`
+
+```python
+class SyncCommitteeContribution(Container):
+    # Slot to which this contribution pertains
+    slot: Slot
+    # Block root for this contribution
+    beacon_block_root: Root
+    # The subcommittee this contribution pertains to out of the broader sync committee
+    subcommittee_index: uint64
+    # A bit is set if a signature from the validator at the corresponding
+    # index in the subcommittee is present in the aggregate `signature`.
+    aggregation_bits: Bitvector[SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT]
+    # Signature by the validator(s) over the block root of `slot`
+    signature: BLSSignature
+```
+
+### `ContributionAndProof`
+
+```python
+class ContributionAndProof(Container):
+    aggregator_index: ValidatorIndex
+    contribution: SyncCommitteeContribution
+    selection_proof: BLSSignature
+```
+
+### `SignedContributionAndProof`
+
+```python
+class SignedContributionAndProof(Container):
+    message: ContributionAndProof
+    signature: BLSSignature
+```
+
+## Validator assignments
+
+A validator determines beacon committee assignments and beacon block proposal duties as defined in the Phase 0 document.
+
+### Sync Committee
+
+To determine sync committee assignments, a validator can run the following function: `is_assigned_to_sync_committee(state, epoch, validator_index)` where `epoch` is an epoch number within the current or next sync committee period.
+This function is a predicate indicating the presence or absence of the validator in the corresponding sync committee for the queried sync committee period.
+
+```python
+def compute_sync_committee_period(epoch: Epoch) -> uint64:
+    return epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD
```

Comment:
What if the fork epoch does not align with the sync committee period? Will we have a shorter period at the start?

Reply:
that seems to be the case, don't think it is worth extra complexity to adjust in protocol. if we have security concerns here we can add <24hrs of epochs to the fork epoch and just wait until it is aligned